pkg/loadbalancer/experimental/benchmark_test.go

```diff
@@ -0,0 +1,222 @@
+
+	name := loadbalancer.ServiceName{Namespace: "test", Name: "svc"}
+	wtxn := p.Services.WriteTxn()
+	p.Services.UpsertFrontend(
+		wtxn,
+		&experimental.FrontendParams{
+			Name:    name,
+			Address: *loadbalancer.NewL3n4Addr(loadbalancer.TCP, addrCluster1, 12345, loadbalancer.ScopeExternal),
+			Type:    loadbalancer.SVCTypeClusterIP,
+			Source:  source.Kubernetes,
+		},
+	)
+	wtxn.Commit()
+
+	numObjects := 1000
+
+	// Add 'numObjects' existing objects to the table.
+	wtxn = p.Services.WriteTxn()
+	for i := 0; i < numObjects; i++ {
+		beAddr := *loadbalancer.NewL3n4Addr(loadbalancer.TCP, addrCluster1, uint16(i), loadbalancer.ScopeExternal)
+		p.Services.UpsertBackends(
+			wtxn,
+			name,
+			source.Kubernetes,
+			&loadbalancer.Backend{
+				L3n4Addr: beAddr,
+				State:    loadbalancer.BackendStateActive,
+			},
+		)
+	}
+	wtxn.Abort()
+
+	// Benchmark the speed at which a new backend is upserted. 'numObjects' are inserted in one
+	// WriteTxn to amortize the cost of WriteTxn&Commit.
+	for n := 0; n < b.N; n++ {
+		wtxn = p.Services.WriteTxn()
+		for i := 0; i < numObjects; i++ {
+			beAddr := *loadbalancer.NewL3n4Addr(loadbalancer.TCP, addrCluster2, uint16(i), loadbalancer.ScopeExternal)
+			p.Services.UpsertBackends(
+				wtxn,
+				name,
+				source.Kubernetes,
+				&loadbalancer.Backend{
+					L3n4Addr: beAddr,
+					State:    loadbalancer.BackendStateActive,
+				},
+			)
+		}
+		// Don't commit the changes so we actually test the cost of Insert() of new object.
+		wtxn.Abort()
+	}
+
+	b.StopTimer()
+	b.ReportMetric(float64(b.N*numObjects)/b.Elapsed().Seconds(), "objects/sec")
+}
+func BenchmarkReplaceBackend(b *testing.B) {
+	p := servicesFixture(b)
+
+	b.ResetTimer()
+
+	addrCluster1 := types.MustParseAddrCluster("1.0.0.1")
+	addrCluster2 := types.MustParseAddrCluster("2.0.0.2")
+
+	name := loadbalancer.ServiceName{Namespace: "test", Name: "svc"}
+	wtxn := p.Services.WriteTxn()
+	p.Services.UpsertFrontend(
+		wtxn,
+		&experimental.FrontendParams{
+			Name:    name,
+			Address: *loadbalancer.NewL3n4Addr(loadbalancer.TCP, addrCluster1, 12345, loadbalancer.ScopeExternal),
+			Type:    loadbalancer.SVCTypeClusterIP,
+			Source:  source.Kubernetes,
+		},
+	)
+
+	beAddr := *loadbalancer.NewL3n4Addr(loadbalancer.TCP, addrCluster2, uint16(1234), loadbalancer.ScopeExternal)
+	p.Services.UpsertBackends(
+		wtxn,
+		name,
+		source.Kubernetes,
+		&loadbalancer.Backend{
+			L3n4Addr: beAddr,
+			State:    loadbalancer.BackendStateActive,
+		},
+	)
+	wtxn.Commit()
+
+	wtxn = p.Services.WriteTxn()
+	b.ResetTimer()
+	for i := 0; i < b.N; i++ {
+		p.Services.UpsertBackends(
+			wtxn,
+			name,
+			source.Kubernetes,
+			&loadbalancer.Backend{
+				L3n4Addr: beAddr,
+				State:    loadbalancer.BackendStateActive,
+			},
+		)
+	}
+	wtxn.Abort()
+
+	b.StopTimer()
+	b.ReportMetric(float64(b.N)/b.Elapsed().Seconds(), "objects/sec")
+}
+
+func BenchmarkReplaceService(b *testing.B) {
+	p := servicesFixture(b)
+
+	b.ResetTimer()
```

Comment:
The timer is reset in line 203, so this is effectively a NoOp and can be removed.